### PR TITLE
OIDC ID Token, Authorization Headers, Refreshing and Verification

### PR DESCRIPTION
**Pull request by JoelSpeed**
This PR replaces: https://github.com/bitly/oauth2_proxy/pull/621

## Description

This PR includes the commits from/replaces #534 and #620 and adds proper OIDC token validation. Please see both original PRs for descriptions and conversations about the PRs.

This now fixes #530

To summarise:

#534: Adds the ability to expose the raw OIDC token to applications behind the proxy
#620: Implement background refreshing of OIDC tokens when cookie_refresh is set > 0
Verification: Implement ValidateSessionState for the OIDC provider to properly verify the id_token. The default providers implementation does not work with OIDC and was causing issues with refreshing

This PR has become quite large but, having opened #620, I realised there was a bug if the cookie_refresh period was shorter than the id_token lifetime (Thanks @jhohertz), to fix this, I needed to correctly implement ValidateSessionState for OIDC which then in turn needs the changes from #534

So the resulting PR is a combination of fixes that adds a bunch of new stuff to the OIDC provider

## Motivation and Context

This makes the OIDC implementation complete in terms of refreshing and token verification. It also allows the OIDC ID tokens to be exposed via an authorization header which can then be proxied into the Kubernetes Dashboard.

## How Has This Been Tested?

This has been running in our production for around 6 months now and we haven't seen any issues with it. OAuth2 Proxy plus dex with the OAuth2 Proxy in nginx auth_request mode

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -212,6 +212,7 @@ Usage of oauth2_proxy:
   -https-address string: <addr>:<port> to listen on for HTTPS clients (default ":443")
   -login-url string: Authentication endpoint
   -pass-access-token: pass OAuth access_token to upstream via X-Forwarded-Access-Token header
+  -pass-authorization-header: pass OIDC IDToken to upstrea via Authorization Bearer header
```
Comment (icelynjennings) [inline README.md]:
```suggestion
  -pass-authorization-header: pass OIDC IDToken to upstream via Authorization Bearer header
```

File: oauthproxy.go [hunk 2]
```diff
@@ -278,15 +282,92 @@ func (p *OAuthProxy) redeemCode(host, code string) (s *providers.SessionState, e
...
+		RawExpires: c.RawExpires,
+		MaxAge:     c.MaxAge,
+		Secure:     c.Secure,
+		HttpOnly:   c.HttpOnly,
+		Raw:        c.Raw,
+		Unparsed:   c.Unparsed,
+	}
+}
+
+func splitCookie(c *http.Cookie) []*http.Cookie {
```
Comment (syscll) [inline oauthproxy.go]: Can we comment this function.

File: oauthproxy.go [hunk 3]
```diff
@@ -278,15 +282,92 @@ func (p *OAuthProxy) redeemCode(host, code string) (s *providers.SessionState, e
...
+			newValue := valueBytes[:3840]
+			valueBytes = valueBytes[3840:]
+			new.Value = string(newValue)
+		}
+		cookies = append(cookies, new)
+	}
+	return cookies
+}
+
+func joinCookies(cookies []*http.Cookie) (*http.Cookie, error) {
```
Comment (syscll) [inline oauthproxy.go]: Can we comment this function.

File: oauthproxy.go [hunk 4]
```diff
@@ -278,15 +282,92 @@ func (p *OAuthProxy) redeemCode(host, code string) (s *providers.SessionState, e
...
+	}
+	c := copyCookie(cookies[0])
+	for i := 1; i < len(cookies); i++ {
+		c.Value += cookies[i].Value
+	}
+	c.Name = strings.TrimRight(c.Name, "-0")
+	return c, nil
+}
+
+func loadCookie(req *http.Request, cookieName string) (*http.Cookie, error) {
```
Comment (syscll) [inline oauthproxy.go]: Can we comment this function.

File: oauthproxy.go [hunk 5]
```diff
@@ -319,6 +400,7 @@ func (p *OAuthProxy) makeCookie(req *http.Request, name string, value string, ex
 // ClearCSRFCookie creates a cookie to unset the CSRF cookie stored in the user's
 // session
 func (p *OAuthProxy) ClearCSRFCookie(rw http.ResponseWriter, req *http.Request) {
+
```
Comment (syscll) [inline oauthproxy.go]: Remove new line.

**Comment by luispollo** [comment]
Hello @JoelSpeed. Thanks for contributing this! Do you have a Docker image built from this PR that is publicly accessible, like you did for the previous repo?

**Comment by jhohertz** [comment]
Just wanted to jump in and day we've been using some of these patches from @JoelSpeed  for around 6 months as well with great results, and am excited to see some new life in the project on his repositories.

Let me know anything I can do to help move things forward.

**Comment by luispollo** [comment]
@JoelSpeed Never mind, I see the Dockerfile in the repo. 😝

File: README.md (hunk 1 above)
Comment (JoelSpeed) [reply README.md]: Fixed

File: oauthproxy.go (hunk 2 above)
Comment (JoelSpeed) [reply oauthproxy.go]: Fixed

File: oauthproxy.go (hunk 3 above)
Comment (JoelSpeed) [reply oauthproxy.go]: Fixed

File: oauthproxy.go (hunk 4 above)
Comment (JoelSpeed) [reply oauthproxy.go]: Fixed

**Comment by tlawrie** [comment]
@JoelSpeed the refresh token is working however I am experiencing the issue where by multiple requests made at the same time (due to reactjs calls) are causing a failure.

One of the calls will make it through to refresh the token, however the other duplicate calls will fail with

```
createSessionState() - rawIDToken: 
createSessionState() - ok: false
RefreshSessionIfNeeded() - Error: unable to update session: token response did not contain an id_token
```

Which means I have multiple calls happening at the same time, and one call succeeds with the refresh. The rests make the refresh call to not get a good response as the refresh token has already been redeemed.

I implemented what is suggested here (https://gist.github.com/JoelSpeed/9f4dbf6f79f6498d12ccd6ff0bc096e2) lines 41-53 as you have previously mentioned back last year.

**Comment by JoelSpeed** [comment]
@tlawrie We've been running that fix for a while now and have seen the problem pretty much disappear. It would be good to think up a long term proper solution for this but the only way I can see that happening is by having some central state storage which would defeat the lovely horizontal scalability of the current design. Welcome to ideas on that!

**Comment by tlawrie** [comment]
@JoelSpeed it might be my implementation of the fix. I currently have the following in the ingress and ingress-controller

**Ingress**
```
"ingress.kubernetes.io/configuration-snippet": "
    proxy_pass_request_body     off;
    proxy_cache                   authentication;
    proxy_cache_key               $cookie_bmrg_auth_proxy_dev;
    proxy_cache_valid             202 401 3s;
    proxy_cache_lock              on;
    proxy_ignore_headers          Set-Cookie;",
```

**Ingress Controller**
```
"http-snippet": "proxy_cache_path        /var/run/cache levels=1:2 keys_zone=authentication:10m inactive=3s;\n",
```

It could also be that something strange is happening in the oidc provider when connected to our provider that is causing the empty ID Token. Looking at the code in RedeemRefreshToken() it merges the context with the new refreshtoken etc. I am not sure how the context.Background() works. Is it possible that it can have an empty ID Token?

File: oauthproxy.go [hunk 6]
```diff
@@ -750,6 +847,12 @@ func (p *OAuthProxy) Authenticate(rw http.ResponseWriter, req *http.Request) int
 	if p.PassAccessToken && session.AccessToken != "" {
 		req.Header["X-Forwarded-Access-Token"] = []string{session.AccessToken}
 	}
+	if p.PassAuthorization && session.IDToken != "" {
+		req.Header["Authorization"] = []string{fmt.Sprintf("Bearer %s", session.IDToken)}
```
Comment (tamalsaha) [inline oauthproxy.go]: @JoelSpeed , I wonder why the idToken is set as Authorization header? Should not this be set as X-Auth-Id-Token or something like that?

**Pull request merged**